include/nbl/video/IGPUCommandBuffer.h

```diff
@@ -67,60 +66,90 @@ class NBL_API IGPUCommandBuffer :
+    bool bindVertexBuffers(uint32_t firstBinding, uint32_t bindingCount, const buffer_t* const* const pBuffers, const size_t* pOffsets) override final;
+    bool dispatchIndirect(const buffer_t* buffer, size_t offset) override final;
+    bool setEvent(event_t* _event, const SDependencyInfo& depInfo) override final;
+    bool resetEvent(event_t* _event, asset::E_PIPELINE_STAGE_FLAGS stageMask) override final;
+    bool waitEvents(uint32_t eventCount, event_t* const* const pEvents, const SDependencyInfo* depInfos) override final;
+    bool drawMeshBuffer(const meshbuffer_t* meshBuffer) override final;
+    bool copyBuffer(const buffer_t* srcBuffer, buffer_t* dstBuffer, uint32_t regionCount, const asset::SBufferCopy* pRegions) override final;
+    bool copyImage(const image_t* srcImage, asset::IImage::E_LAYOUT srcImageLayout, image_t* dstImage, asset::IImage::E_LAYOUT dstImageLayout, uint32_t regionCount, const asset::IImage::SImageCopy* pRegions) override final;
+    bool copyBufferToImage(const buffer_t* srcBuffer, image_t* dstImage, asset::IImage::E_LAYOUT dstImageLayout, uint32_t regionCount, const asset::IImage::SBufferCopy* pRegions) override final;
+    bool blitImage(const image_t* srcImage, asset::IImage::E_LAYOUT srcImageLayout, image_t* dstImage, asset::IImage::E_LAYOUT dstImageLayout, uint32_t regionCount, const asset::SImageBlit* pRegions, asset::ISampler::E_TEXTURE_FILTER filter) final override;
+    bool copyImageToBuffer(const image_t* srcImage, asset::IImage::E_LAYOUT srcImageLayout, buffer_t* dstBuffer, uint32_t regionCount, const asset::IImage::SBufferCopy* pRegions) override final;
+    bool resolveImage(const image_t* srcImage, asset::IImage::E_LAYOUT srcImageLayout, image_t* dstImage, asset::IImage::E_LAYOUT dstImageLayout, uint32_t regionCount, const asset::SImageResolve* pRegions) override final;
+    bool executeCommands(uint32_t count, cmdbuf_t* const* const cmdbufs) override final;
+    bool regenerateMipmaps(IGPUImage* img, uint32_t lastReadyMip, asset::IImage::E_ASPECT_FLAGS aspect) override final;
 
-    SInheritanceInfo getCachedInheritanceInfo() const
-    {
-        return m_cachedInheritanceInfo;
-    }
+    inline uint32_t getQueueFamilyIndex() const { return m_cmdpool->getQueueFamilyIndex(); }
+    inline IGPUCommandPool* getPool() const { return m_cmdpool.get(); }
+    inline SInheritanceInfo getCachedInheritanceInfo() const { return m_cachedInheritanceInfo; }
 
     // OpenGL: nullptr, because commandbuffer doesn't exist in GL (we might expose the linked list command storage in the future)
     // Vulkan: const VkCommandBuffer*
     virtual const void* getNativeHandle() const = 0;
 
-protected:
+    inline const core::unordered_map<const IGPUDescriptorSet*, uint64_t>& getBoundDescriptorSetsRecord() const { return m_boundDescriptorSetsRecord; }
+
+    inline void clearBoundDescriptorSetsRecord() { m_boundDescriptorSetsRecord.clear(); }
```

Comment:
why are these public?

Reply:
just forward declare `IGPUQueue` and make it a friend (and put the friendship declaration near the variable it was required for)

Reply:
the only time you should clear this is upon a `reset` (which includes a `begin()`) or a detected pool reset

Reply:
hmm maybe only `checkForParentPoolReset` and `releaseResourcesBackToPool`

Reply:
> furthermore clearBoundDescriptorSetsRecord should not exist at all, think... what happens if you re-submit the same commandbuffer multiple times? You're supposed to check every time that the descriptor sets were not updated.

Right, this makes sense.

Reply:
> hmm maybe only checkForParentPoolReset and releaseResourcesBackToPool

Yeah I think that will work.